ucb-rit/coldfront
# Add support for user requests to join projects

Fixes #26

**Changes**
- Added a field `joins_require_approval` to `Project` that stores whether or not project managers need to approve requests to join their project.
- Added views to search for a project to join and to request to join one.
- Added a view for project managers to approve or deny join requests.
- Modified the view for updating project information to allow changes to the `joins_require_approval` setting.
    - By default, all requests are approved, but managers can change this.
    - If the setting is changed to `False` (meaning approval is not required), all pending requests are approved automatically.

**How to Test**
- Review the code changes and add comments as needed.
- Ensure that the test suite passes.
- Manual Testing
    1. Default Automatic Approval
        - As a non-PI user, navigate to "Projects" and then "Join a project".
            - Projects on which the current user is already a member should have the "Join" button greyed out.
            - Join a project.
        - As a PI user on that project, navigate to the Project detail view.
            - By default, all join requests are approved, so the user should now be a member.
            - Delete the newly added member.
    2. Approval Required
        - As the PI user, update the setting for `joins_require_approval` so that approval is required. On refresh, a "Review Join Requests" button should now appear.
        - As the non-PI user, make another request.
        - As the PI user, approve/deny the request. The user should/should not be added.
        - Delete the newly added member.
    3. Setting Change
        - As the PI user, ensure that approval is required.
        - As the non-PI user, make another request.
        - As the PI user, change the setting so that approval is not required.
            - The request should be approved automatically.

**Notes**
- Email handling is not covered by this pull request.
